### ISSUE TITLE
Define genmat for AbstractHyperrectangle

### ISSUE THREAD

**Issue by schillic** [JuliaReach/LazySets.jl#1878]
Currently we use the fallback implementation

https://github.com/JuliaReach/LazySets.jl/blob/85e4b30770a2d888a1b10951ed687e1ac49accbd/src/Interfaces/AbstractHyperrectangle.jl#L59-L61

which is

https://github.com/JuliaReach/LazySets.jl/blob/85e4b30770a2d888a1b10951ed687e1ac49accbd/src/Interfaces/AbstractZonotope.jl#L68-L74

However, in another place we use

https://github.com/JuliaReach/LazySets.jl/blob/85e4b30770a2d888a1b10951ed687e1ac49accbd/src/convert.jl#L258-L260

which seems more efficient.

**Comment by mforets** [comment]
I agree, `genmat(H::AbstractHyperrectangle)` should produce a `Diagonal` matrix.

**Comment by schillic** [comment]
> I agree, `genmat(H::AbstractHyperrectangle)` should produce a `Diagonal` matrix.

I implemented this in #1880, but that breaks a test: in flat dimensions the `Diagonal` contains redundant columns. I will now do the following: If there is a flat dimension, I will return a sparse matrix; otherwise I will return the `Diagonal`.

**Comment by mforets** [comment]
The genma of an hyperrectangular is always diagonal. Redundant entries in `radius_hyperrectangle` can be removed in several ways, eg. using `Base.unique`.

**Comment by mforets** [comment]
Ok, you would like to remove the zero columns in the `Diagonal` IIUC. The intention of returning either sparse or diagonal is good, but it is bad for performance. I would always return a `Diagonal`, such that the type of the output can be inferred from the type of the input. If there is not yet a function, we can add `remove_zero_generators` to remove zero generators.

**Comment by schillic** [comment]
> I would always return a Diagonal, such that the type of the output can be inferred from the type of the input.

As I tried to explain, this is not possible. I can offer the compromise to always return a `SparseMatrix`.

**Comment by mforets** [comment]
Can you elaborate? Do you have an example?

**Comment by schillic** [comment]
See the discussion in https://github.com/JuliaReach/LazySets.jl/pull/1880#discussion_r361808082.
